src/main/flight/pid.c

```diff
@@ -803,77 +802,92 @@ void FAST_CODE pidController(const pidProfile_t *pidProfile, const rollAndPitchT
-
-        // -----calculate P component and add Dynamic Part based on stick input
-        pidData[axis].P = pidCoefficient[axis].Kp * errorRate * tpaFactor;
-        if (axis == FD_YAW) {
-            pidData[axis].P = ptermYawLowpassApplyFn((filter_t *) &ptermYawLowpass, pidData[axis].P);
-        }
 
-        // -----calculate I component
-        float itermErrorRate = 0.0f;
-#if defined(USE_ABSOLUTE_CONTROL)
+#ifdef USE_ABSOLUTE_CONTROL
+        float acCorrection = 0;
         float acErrorRate;
-#endif
+#endif        
+        float itermErrorRate = 0.0f;
+
 #if defined(USE_ITERM_RELAX)
         if (itermRelax && (axis < FD_YAW || itermRelax == ITERM_RELAX_RPY )) {
-            float gyroTargetLow = pt1FilterApply(&windupLpf[axis][0], currentPidSetpoint);
-            float gyroTargetHigh = pt1FilterApply(&windupLpf[axis][1], currentPidSetpoint);
-#if defined(USE_ABSOLUTE_CONTROL)
-            gyroTargetLow += acCorrection;
-            gyroTargetHigh += acCorrection;
-#endif
-            if (axis < FD_YAW) {
-                int itemOffset = (axis << 1);
-                DEBUG_SET(DEBUG_ITERM_RELAX, itemOffset++, gyroTargetHigh);
-                DEBUG_SET(DEBUG_ITERM_RELAX, itemOffset, gyroTargetLow);
+            const float gyroTarget = pt1FilterApply(&windupLpf[axis], currentPidSetpoint);
+            const float gyroHpf = fabsf(currentPidSetpoint - gyroTarget);
+            if (itermRelaxType == ITERM_RELAX_SETPOINT) {
+                itermErrorRate = (1 - MIN(1, fabsf(gyroHpf) / 60.0f)) * (currentPidSetpoint - gyroRate);
+            }
+            const float tolerance = gyroHpf * 1.0f;
+            if (axis == FD_ROLL) {
+                DEBUG_SET(DEBUG_ITERM_RELAX, 0, gyroTarget);
+                DEBUG_SET(DEBUG_ITERM_RELAX, 1, gyroTarget + tolerance);
+                DEBUG_SET(DEBUG_ITERM_RELAX, 2, gyroTarget - tolerance);
+                DEBUG_SET(DEBUG_ITERM_RELAX, 3, axisError[axis] * 10);
             }
-            const float gmax = MAX(gyroTargetHigh, gyroTargetLow);
-            const float gmin = MIN(gyroTargetHigh, gyroTargetLow);
-            if (gyroRate < gmin || gyroRate > gmax) {
-                itermErrorRate = (gyroRate > gmax ? gmax : gmin ) - gyroRate;
+            if (itermRelaxType == ITERM_RELAX_GYRO ) {
+                const float gmax = gyroTarget + tolerance;
+                const float gmin = gyroTarget - tolerance;
+                
+                if (gyroRate >= gmin && gyroRate <= gmax) {
+                     itermErrorRate = 0;
+                } else {
+                     itermErrorRate = (gyroRate > gmax ? gmax : gmin ) - gyroRate;
+                }
             }
+            
 #if defined(USE_ABSOLUTE_CONTROL)
-            else {
-                itermErrorRate = acCorrection;
+            const float gmaxac = gyroTarget + 2 * tolerance;
+            const float gminac = gyroTarget - 2 * tolerance;
+            if (gyroRate >= gminac && gyroRate <= gmaxac) {
```

Comment:
following code is equivalent to (but I don't quite understand it's meaning, there is discontinuity at +-2* tolerance ):
```
if(fabs(gyroTarget - gyroRate) <= 2 * tolerance) {
    acErrorRate = (gyroTarget - gyroRate) + 2 * tolerance * ( axisError[axis] < 0 ? 1 : -1); // or -fsign() etc.
    acErrorRate = constrainf(acErrorRate, -axisError[axis]/dT, axisError[axis]/dT);
} else {
  acErrorRate = fapplyDeadband(gyroTarget - gyroRate, 2* tolerance);
}

Reply:
Leaving this as it is right now, pending clarification of what it does.

Reply:
This is to allow the the accumulated acError to reduce quickly when in the target path. We want error only to be added when out side the target band. But reduced when within, not matter at which position in the band.